python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -85,75 +83,151 @@ def __init__(
         ):
             self._reader_registered = True
 
+        self._meta_channel: Optional[Channel] = _meta_channel
+        if (
+            self._meta_channel is None
+            and self._writer_registered
+            and (
+                self._typ.shape == TorchTensorType.AUTO
+                or self._typ.dtype == TorchTensorType.AUTO
+            )
+        ):
+            # Allocate 1KiB for metadata.
+            metadata_type = SharedMemoryType(buffer_size_bytes=1_000)
+            self._meta_channel = metadata_type.create_channel(
+                self._writer,
+                self._readers,
+            )
+
+        if self._meta_channel is None:
+            # If there is no metadata channel, then we can only pass tensors of
+            # static shape and dtype.
+            assert self.has_static_type()
+
     def ensure_registered_as_writer(self):
         assert self._nccl_group is not None, "Actor is not part of a NCCL group"
+        assert self._writer_registered
+        assert self._device.type == "cuda"
 
     def ensure_registered_as_reader(self) -> bool:
         assert self._nccl_group is not None, "Actor is not part of a NCCL group"
+        assert self._reader_registered
+        assert self._device.type == "cuda"
 
     def __reduce__(self):
-        return (self.__class__, (self._writer, self._readers, self._typ, self._device))
+        return (
+            self.__class__,
+            (self._writer, self._readers, self._typ, self._meta_channel),
+        )
 
-    def write(
-        self,
-        wrapped: "_TorchTensorWrapper",
-        readers: Optional[List[ray.actor.ActorHandle]] = None,
-    ):
-        value = wrapped.tensor
-        if value.shape != self._typ.shape:
+    def _get_tensor_meta(self, tensor: "torch.Tensor") -> Optional["TorchTensorType"]:
+        if not isinstance(tensor, self.torch.Tensor):
+            raise ValueError("Task must return torch.Tensors")
+
+        if tensor.device != self._device:
             raise ValueError(
-                f"torch.Tensor has shape {value.shape}, expected {self._typ.shape}"
+                f"torch.Tensor must be on the default device: {self._device}"
             )
-        if value.dtype != self._typ.dtype:
+
+        meta: Optional["TorchTensorType"] = None
+        if (
+            self._typ.shape == self.TorchTensorType.AUTO
+            or self._typ.dtype == self.TorchTensorType.AUTO
+        ):
+            meta = self.TorchTensorType(shape=tensor.shape, dtype=tensor.dtype)
```

Comment:
self.TorchTensorType(shape=tensor.shape, dtype=tensor.dtype) -> this code is a bit weird. Is this a valid syntax? Isn't it just `TorchTensorType(shape=tensor.shape, dtype=tensor.dtype)`?

Reply:
Ah yeah I was just trying to avoid all of the ugly imports, but I'll just change it to use them.